### ISSUE TITLE
What is the recommended way of training GANs?

### ISSUE THREAD

**Issue by yuxinyuan** [huggingface/accelerate#174]
Currently, the examples folder doesn't contain any example of training GAN. I wonder what is the recommended way of handling multiple models and optimizers when using accelerate.

In terms of interface, `Accelerator.prepare` can wrap arbitrary number of models and optimizers at once. However, it seems to me that the current implementation of `Accelerator` only has one gradient scaler (when native amp is enabled), this might potentially cause issues for wrapping multiple optimizers with one `Accelerator` instance. One way of fixing this might be to move the ownership of `GradScaler` to an `AcceleratedOptimizer`, but this would case problems when calling `Accelerator.backward`.

On the other hand, to use deepspeed, one would have to create two `DeepSpeedEngine` instances to wrap two models. Maybe accelerate could follow this pattern, since deepspeed would be one of the supported backend.

Anyway, I guess a minimum GAN training script should be added to examples as a guildline.

**Comment by sgugger** [comment]
Hi there! There is no example of GAN training yet in the library though it would be super interesting to have one and see where the API limits are hit and how they should be adapted to be fixed. If you or anyone want to contribute such an example, that would be very valuable.

**Comment by EyalMichaeli** [comment]
I'd love to see such an example

**Comment by gau-nernst** [comment]
I'm using Accelerate to train some GANs with mixed precision. With mixed precision enabled (fp16), I encountered a lot of instabilities pretty early on during the training (NaN loss). Turning off mixed precision solves the problem.

According to PyTorch docs, `scaler.update()` should only be called once at each iteration: https://pytorch.org/docs/stable/notes/amp_examples.html#working-with-multiple-models-losses-and-optimizers

I believe the current implementation of Accelerate will always call `scaler.update()` whenever we call `AcceleratedOptimizer.step()`.

I don't know much about mixed precision training logic, so I'm not sure what will be the correct fix here. I will try the following to see if it fixes the problem
- use bf16 with Accelerate
- use PyTorch AMP directly with only one `scaler.update()` per iteration like described above

**Comment by yuxinyuan** [comment]
@gau-nernst I think you are right that `` Accelerate will always call scaler.update() whenever we call AcceleratedOptimizer.step()``, which violates the guidelines in pytorch doc.

But there's also something else you can do to debug your fp16 training. For example, 1e-8 will be rounded to 0. in fp16, so if you have something like ``torch.log(loss + 1e-8)`` or `` tensora / (tensorb + 1e-8)``. These could cause NaN loss in fp16 training.

**Comment by gau-nernst** [comment]
@yuxinyuan Thank you for the pointer! Also, from what I understand, mixed-precision only happens inside the model (after I pass `model = accelerator.prepare(model)`. The output is casted to fp32. Therefore, any computation done outside of the model is still in fp32? e.g. loss calculations...

If that's the case, I suppose I only need to debug what's happening inside my model? The computations outside the model should be in full precision so it should be fine?

**Comment by johnsutor** [comment]
Great demo of how to do so here: https://github.com/huggingface/community-events/blob/af4c17ebd748398e40e15ca10282d502fb69c757/huggan/pytorch/lightweight_gan/lightweight_gan.py#L567
